chain-tx-enclave-next/mls/src/group.rs

```diff
@@ -168,42 +217,76 @@ impl GroupAux {
         }
     }
 
-    fn init_commit(&mut self, add_proposals: &[MLSPlaintext]) -> (MLSPlaintext, Welcome) {
-        let add_proposals_ids: Vec<ProposalId> = add_proposals
-            .iter()
-            .map(|plain| ProposalId(self.tree.cs.hash(&plain.get_encoding())))
-            .collect();
+    fn commit_proposals(&mut self, proposals: &[MLSPlaintext]) -> (MLSPlaintext, Welcome) {
+        // split proposals by types
+        let mut add_proposals_ids: Vec<ProposalId> = Vec::new();
+        let mut additions: Vec<Add> = Vec::new();
+        let mut update_proposals_ids: Vec<ProposalId> = Vec::new();
+        let mut updates: Vec<(LeafSize, Update)> = Vec::new();
+        let mut remove_proposals_ids: Vec<ProposalId> = Vec::new();
+        let mut removes: Vec<Remove> = Vec::new();
+        for p in proposals.iter() {
+            let proposal_id = ProposalId(self.tree.cs.hash(&p.get_encoding()));
+            match &p.content.content {
+                ContentType::Proposal(Proposal::Add(add)) => {
+                    add_proposals_ids.push(proposal_id);
+                    additions.push(add.clone());
+                }
+                ContentType::Proposal(Proposal::Update(update)) => {
+                    update_proposals_ids.push(proposal_id);
+                    updates.push((LeafSize(p.content.sender.sender as usize), update.clone()));
+                }
+                ContentType::Proposal(Proposal::Remove(remove)) => {
+                    remove_proposals_ids.push(proposal_id);
+                    removes.push(remove.clone());
+                }
+                _ => panic!("invalid proposal message type"),
+            }
+        }
         let mut updated_tree = self.tree.clone();
-        let positions = updated_tree.update(&add_proposals, &[], &[]);
+        let positions = updated_tree.update(&additions, &updates, &removes);
 
-        // generate init key
-        self.owned_kp.update_init_key();
+        // pathRequired = haveUpdateOrRemove || haveNoProposalsAtAll
+        let should_populate_path =
+            proposals.is_empty() || !updates.is_empty() || !removes.is_empty();
```

Comment:
the fact that the path must be populated in these cases does not imply that in the other cases, it must not be populated.
In particular for the group initialisation / at genesis, it must be populated, as there's no "init_secret" (so no entropy from there, as one just passes empty or 0 vec) and we don't use PSK or something like that.

so either dedicated internal function `init_commit` or some extra parameter to this one?

Reply:
> ## a few clippy suggestions: https://travis-ci.org/github/crypto-com/chain/jobs/702102278#L1710
> I guess for that `(Vec<(NodeSize, KeyPackage)>, Option<SecretVec<u8>>)`, it could either be a type alias with a comment explaining what these are, or struct with fields with meaningful names.
> looks ok, the only thing is the "genesis" initialisation -- I suppose we'd want to populate the path in those first "static" messages that would be put in genesis?

It seems the first part of the tuple is not used yet, I removed it.

Reply:
> the fact that the path must be populated in these cases does not imply that in the other cases, it must not be populated.
> In particular for the group initialisation / at genesis, it must be populated, as there's no "init_secret" (so no entropy from there, as one just passes empty or 0 vec) and we don't use PSK or something like that.
> 
> so either dedicated internal function `init_commit` or some extra parameter to this one?

it seems the commit message generated from genesis initialization is not used by anyone? other nodes are initialized with the welcome message?

Reply:
the group creator would self-apply it :)

Reply:
If the path isn't populated at genesis, the group creator would need to also immediately create another empty Commit distributed along with Welcome

Reply:
(but I guess it's easier if it's just populated in the first commit?)

Reply:
> (but I guess it's easier if it's just populated in the first commit?)

Yes, I agree.
Currently I got another testing failure, which is caused by self-update proposal.
We need to update the parent hash extension of the keypackage included in update proposal when committing, which is seems a bit hacky, not sure is it correct approach.